partition-alloc/src/allocator_shim_default_dispatch_to_partition_alloc.cc

```diff
@@ -0,0 +1,679 @@
+  // feature-test.
+#if !defined(__cpp_lib_atomic_value_initialization) || \
+    __cpp_lib_atomic_value_initialization < 201911L
+  alignas(T) uint8_t instance_buffer_[sizeof(T)];
+#else
+  alignas(T) uint8_t instance_buffer_[sizeof(T)] = {0};
+#endif
+  std::atomic<bool> initialization_lock_;
+};
+
+template <typename T, typename Constructor>
+T* LeakySingleton<T, Constructor>::GetSlowPath() {
+  // The instance has not been set, the proper way to proceed (correct
+  // double-checked locking) is:
+  //
+  // auto* instance = instance_.load(std::memory_order_acquire);
+  // if (!instance) {
+  //   ScopedLock initialization_lock;
+  //   root = instance_.load(std::memory_order_relaxed);
+  //   if (root)
+  //     return root;
+  //   instance = Create new root;
+  //   instance_.store(instance, std::memory_order_release);
+  //   return instance;
+  // }
+  //
+  // However, we don't want to use a base::Lock here, so instead we use
+  // compare-and-exchange on a lock variable, which provides the same
+  // guarantees.
+  SimpleScopedSpinLocker scoped_lock{initialization_lock_};
+
+  T* instance = instance_.load(std::memory_order_relaxed);
+  // Someone beat us.
+  if (instance) {
+    return instance;
+  }
+
+  instance = Constructor::New(reinterpret_cast<void*>(instance_buffer_));
+  instance_.store(instance, std::memory_order_release);
+
+  return instance;
+}
+
+class MainPartitionConstructor {
+ public:
+  static partition_alloc::PartitionRoot* New(void* buffer) {
+    size_t pkey = ::ia2_get_pkey();
+    auto* new_root = new (buffer)
+        partition_alloc::PartitionRoot(partition_alloc::PartitionOptions{
+            .aligned_alloc =
+                partition_alloc::PartitionOptions::AlignedAlloc::kAllowed,
+            .thread_cache = partition_alloc::PartitionOptions::ThreadCache::kDisabled,
+            .backup_ref_ptr =
+                partition_alloc::PartitionOptions::BackupRefPtr::kDisabled,
+            .thread_isolation = partition_alloc::ThreadIsolationOption(pkey, pkey),
+        });
+
+    return new_root;
+  }
+};
+
+LeakySingleton<partition_alloc::PartitionRoot, MainPartitionConstructor>
+    g_root[partition_alloc::internal::kNumCompartments] PA_CONSTINIT = {};
+
+class SharedPartitionConstructor {
+ public:
+  static partition_alloc::PartitionRoot* New(void* buffer) {
+    auto* new_root = new (buffer)
+        partition_alloc::PartitionRoot(partition_alloc::PartitionOptions{
+            .aligned_alloc =
+                partition_alloc::PartitionOptions::AlignedAlloc::kAllowed,
+            .thread_cache = partition_alloc::PartitionOptions::ThreadCache::kDisabled,
+            .backup_ref_ptr =
+                partition_alloc::PartitionOptions::BackupRefPtr::kDisabled,
+        });
+
+    return new_root;
+  }
+};
+
+LeakySingleton<partition_alloc::PartitionRoot, SharedPartitionConstructor>
+    g_shared_root PA_CONSTINIT = {};
+
+partition_alloc::PartitionRoot *Allocator() {
+  size_t pkey = ::ia2_get_pkey();
```

Comment:
This returns pkey 0 when the pkru is not one of the 16 expected values. We should fix that, but it can be in a separate change if you want.

Reply:
Is that not the behavior we want here? Early in process initialization we will have the default pkey, not 0.

Reply:
Do we expect allocations during early process init? IIRC the dynamic linker has its own basic allocator for early allocations and I don't think our runtime does any. Regardless I'd expect the behavior for an unknown pkey to be terminating the process, but I guess allowing shared allocations in this case doesn't hurt.

Reply:
I think for now we should allow them, we can always change this later.